lib/SILOptimizer/Utils/ShrinkBorrowScope.cpp

```diff
@@ -100,9 +108,70 @@ class ShrinkBorrowScope {
   }
 
   bool isDeinitBarrier(SILInstruction *instruction) {
-    return users.contains(instruction) || instruction->maySynchronize() ||
-           mayAccessPointer(instruction, value) ||
-           mayLoadWeakOrUnowned(instruction, value);
+    return isBarrierApply(instruction) || instruction->maySynchronize() ||
+           mayAccessPointer(instruction) || mayLoadWeakOrUnowned(instruction);
+  }
+
+  bool canReplaceValueWithBorrowedValue(SILValue value) {
+    while (true) {
+      auto *instruction = value.getDefiningInstruction();
+      if (!instruction)
+        return false;
+      if (auto *cvi = dyn_cast<CopyValueInst>(instruction)) {
+        value = cvi->getOperand();
+        continue;
+      } else if (auto *bbi = dyn_cast<BeginBorrowInst>(instruction)) {
+        if (bbi == this->bbi) {
```

Comment:
Variable shadowing is dangerous. I'd rename the self member to something more verbose.

Reply:
Fixed.